drivers/sms_ser.c

```diff
@@ -0,0 +1,701 @@
+/*
+ * sms_ser.c: code for mono protocol for SMS Brazil UPSes
+ *
+ * Copyright (C) 2023 - Alex W. Baule <alexwbaule@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Reference of the derivative work: riello driver
+ */
+
+#include "config.h" /* must be the first header */
+#include <string.h>
+
+#include "common.h" /* for upsdebugx() etc */
+#include "sms_ser.h"
+#include "main.h"
+#include "serial.h"
+
+#define ENDCHAR '\r'
+
+#define DRIVER_NAME "SMS Brazil UPS driver"
+#define DRIVER_VERSION "1.00"
+
+#define QUERY_SIZE 7
+#define BUFFER_SIZE 18
+#define RESULT_SIZE 18
+#define HUMAN_VALUES 7
+
+static uint16_t bootdelay = DEFAULT_BOOTDELAY;
+static uint8_t bufOut[BUFFER_SIZE];
+static uint8_t bufIn[BUFFER_SIZE];
+static SmsData DeviceData;
+
+/* driver description structure */
+upsdrv_info_t upsdrv_info = {
+    DRIVER_NAME,
+    DRIVER_VERSION,
+    "Alex W. Baulé <alexwbaule@gmail.com>",
+    DRV_BETA,
+    {NULL}};
+
+void sms_parse_features(uint8_t *rawvalues, SmsData *results) {
+    char tbattery[6];
+    char frequency[4];
+
+    memset(results->voltageRange, 0, sizeof(results->voltageRange));
+    memset(results->currentRange, 0, sizeof(results->currentRange));
+    memset(tbattery, 0, sizeof(tbattery));
+    memset(frequency, 0, sizeof(frequency));
+
+    for (int i = 1; i < BUFFER_SIZE - 2; i++) {
```

Comment:
Here and in other loops, I think some of the compiler portability checks would complain about first declaring a variable inside the loop. Clang tends to do so even if compiling for C99 or newer (which are okay with this), when pedantic warnings are enabled... No worry if CI does not complain with default settings though. Becoming "clean" against more and more warnings types is a separate effort :)